lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -313,6 +314,15 @@ def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_optio
         # TODO: decide how to deprecate the old src API behavior
         req_source = collection_req.get('source', None)
         req_signature_sources = collection_req.get('signatures', None)
+        key_diff = frozenset(collection_req.keys()).difference(SUPPORTED_REQUIREMENTS_KEYS)
+        if key_diff:
+            raise AnsibleError(
+                "The keys (%s) are not valid when installing collection "
+                "requirement entries. Be sure to only use the following "
+                "supported keys (%s)."
+                % (', '.join(key_diff), ', '.join(SUPPORTED_REQUIREMENTS_KEYS))
```

Comment:
I would personally sort the sets before joining them so that the error message becomes more deterministic. (It makes it also easier to scan the list of supported keys if they are always in alphabetical order :) .)